src/Storages/MergeTree/MergeTreeIndexSet.cpp

```diff
@@ -85,6 +87,15 @@ void MergeTreeIndexGranuleSet::serializeBinary(WriteBuffer & ostr) const
         serialization->serializeBinaryBulkWithMultipleStreams(column, 0, size(), settings, state);
         serialization->serializeBinaryBulkStateSuffix(settings, state);
     }
+
+    for (size_t i = 0; i < num_columns; ++i)
```

Comment:
let's add tests to ensure backward and forward compatibility. newer version should be able to read old parts, old versions should be able to read new (just inserted by the new version) parts.
for the reference you may use the tests under `tests/integration/test_backward_compatibility`

Reply:
The current implementation does indeed have compatibility issues. 
the maximum and minimum values are recalculated after deserializing the columns, and then they are added as a Range to `set_hyperrectangle`. `set_hyperrectangle` does not participate in serialization and deserializationThis approach avoids serialization compatibility issues.